invoice/src/main/java/org/killbill/billing/invoice/InvoicePluginDispatcher.java

```diff
@@ -167,22 +172,77 @@ private void onCompletionCall(final boolean isSuccess,
         // We clone the original invoice so plugins don't remove/add items
         final Invoice clonedInvoice = (Invoice) ((DefaultInvoice) originalInvoice).clone();
         for (final InvoicePluginApi invoicePlugin : invoicePlugins) {
-            final List<InvoiceItem> items = invoicePlugin.getAdditionalInvoiceItems(clonedInvoice, isDryRun, ImmutableList.<PluginProperty>of(), callContext);
-            if (items != null) {
-                for (final InvoiceItem item : items) {
-                    validateInvoiceItemFromPlugin(item, invoicePlugin);
-                    additionalInvoiceItems.add(item);
+            final List<InvoiceItem> additionalInvoiceItemsForPlugin = invoicePlugin.getAdditionalInvoiceItems(clonedInvoice, isDryRun, ImmutableList.<PluginProperty>of(), callContext);
+            if (additionalInvoiceItemsForPlugin != null) {
+                for (final InvoiceItem additionalInvoiceItem : additionalInvoiceItemsForPlugin) {
+                    final InvoiceItem sanitizedInvoiceItem = validateAndSanitizeInvoiceItemFromPlugin(originalInvoice, additionalInvoiceItem, invoicePlugin);
+                    additionalInvoiceItems.add(sanitizedInvoiceItem);
```

Comment:
Since you are improving this method -- validation piece -- on suggestion would be to change the following as well:

Today we pass `originalInvoice` and return a list of `additionalInvoiceItems` -- those are not really additional items. they are a mixed bags of modified existing and additional, therefore requiring logic on the caller side to sort that out. Maybe we could encapsulate that logic inside that method instead?

Reply:
`originalInvoice` in the non-API use-case is actually a temporary cloned invoice, with a temporary CBA item computed pre- invoice plugins (https://github.com/killbill/killbill/issues/637). Unfortunately the shuffling of that logic doesn't _fully_ help simplify the `processAccountWithLockAndInputTargetDate` implementation. See  15964b0c682fdfdc0bd130e94ac6113a58b3d016.